avogadro/qtplugins/atompairbonds/atompairbonds.cpp

```diff
@@ -0,0 +1,159 @@
+
+static const float ABSOLUTE_MAX_DISTANCE = 2.0;
+
+static int getInteractionType(const Molecule &molecule, Index i) {
+  unsigned char inum = molecule.atomicNumber(i);
+  switch (inum) {
+    case 1: // hydrogen bond
+      for (const Bond *b : molecule.bonds(i)) {
+        Index j = (b->atom1().index() == i ? b->atom2() : b->atom1()).index();
+        unsigned char jnum = molecule.atomicNumber(j);
+        switch (jnum) {
+          case 7: case 8: case 9: // F, O, N
+            return HYDROGEN_BOND;
+        }
+      }
+      break;
+  }
+  return -1;
+}
+
+static bool checkPairDonorIsValid(const Molecule &molecule, Index n, int interactionType) {
+  unsigned char nnum = molecule.atomicNumber(n);
+  switch (interactionType) {
+    case HYDROGEN_BOND:
+      switch (nnum) {
+        case 7: case 8: case 9: // F, O, N
+          return true;
+      }
+      break;
+  }
+  return false;
+}
+
+void AtomPairBonds::process(const Molecule &molecule, Rendering::GroupNode &node)
+{
+  float radius(0.1f);
+  Vector3ub color(64, 192, 255);
+
+  NeighborPerceiver perceiver(molecule.atomPositions3d(), ABSOLUTE_MAX_DISTANCE);
+  std::vector<bool> isAtomEnabled(molecule.atomCount());
+  for (Index i = 0; i < molecule.atomCount(); ++i)
+    isAtomEnabled[i] = m_layerManager.atomEnabled(i);
+
+  GeometryNode *geometry = new GeometryNode;
+  node.addChild(geometry);
+  DashedLineGeometry *lines = new DashedLineGeometry;
+  lines->identifier().molecule = &molecule;
+  lines->identifier().type = Rendering::BondType;
+  lines->setLineWidth(2.0);
+  geometry->addDrawable(lines);
+  Array<Index> neighbors;
+  for (Index i = 0; i < molecule.atomCount(); ++i) {
+    if (!isAtomEnabled[i])
+      continue;
+    int interactionType = getInteractionType(molecule, i);
+    if (interactionType < 0)
+      continue;
+    Vector3 pos = molecule.atomPosition3d(i);
+    perceiver.getNeighborsInclusiveInPlace(neighbors, pos);
+    for (Index n : neighbors) {
+      if (!isAtomEnabled[n])
+        continue;
+      if (!checkPairDonorIsValid(molecule, n, interactionType))
+        continue;
+
+      Vector3 npos = molecule.atomPosition3d(n);
+      Vector3 distance_vector = npos - pos;
+
+      if (distance_vector.norm() > MAX_DISTANCES[interactionType])
+        continue;
+
+      lines->addDashedLine(pos.cast<float>(), npos.cast<float>(), color, 8);
+    }
+  }
+}
+
+QWidget *AtomPairBonds::setupWidget()
+{
+  QWidget *widget = new QWidget(qobject_cast<QWidget *>(this->parent()));
+  QVBoxLayout *v = new QVBoxLayout;
+
+  // maximum distance
+  QDoubleSpinBox *spin = new QDoubleSpinBox;
+  spin->setRange(0.0, 180.0);
+  spin->setSingleStep(1.0);
+  spin->setDecimals(0);
+  spin->setSuffix(tr(" °"));
+  spin->setValue(m_angleToleranceDegrees);
+  QObject::connect(spin, SIGNAL(valueChanged(double)), this,
+                   SLOT(setMaximumDistance(double)));
+  QFormLayout *form = new QFormLayout;
+  //form->addRow(QObject::tr("Angle tolerance:"), spin);
```

Comment:
I'm confused - why don't you need this?

Reply:
It has no effect at the moment and was slightly bugging me. But I guess it doesn't hurt to have it...

Reply:
Well, it has no effect because you've set the `SLOT(setMaximumDistance(double))` which doesn't exist. Presumably you want `setAngleTolerance(double)`

You can use the newer syntax:
```
connect(spin, &valueChanged, this, &setMaximumDistance);
```

This should catch problems at compile time.